aim/agent/aid/universes/aci/aci_universe.py

```diff
@@ -113,37 +124,152 @@ def _reload_websocket_config(self):
+                self.apic_password,
+                verify_ssl=self.verify_ssl_certificate,
+                cert_name=self.cert_name,
+                key=self.private_key_file)
+            resp = session.login()
+            if not resp.ok:
+                LOG.warn('%s Websocket connection failed: %s',
+                         purpose, resp.text)
+                retries += 1
+                if session.session:
+                    session.close()
+                continue
+            LOG.info('%s Websocket connection succeeded with url: %s',
+                     purpose, url)
+            return session
+
+        return None
+
+    def establish_ws_session_for_recovery(self, max_retries=None):
+        try:
+            with utils.get_rlock(lcon.ACI_WS_CONNECTION_LOCK, blocking=False):
+                max_retries = max_retries or 2 * len(self.ws_urls)
+                url_max_retries = max(1, max_retries / len(self.ws_urls))
+                aim_context = aim_ctx.AimContext(store=api.get_store())
+                for url in self.ws_urls:
+                    apic_assign = api_infra.ApicAssignment(apic_host=url)
+                    apic_assign_obj = self.manager.get(aim_context,
+                                                       apic_assign)
+                    if (apic_assign_obj and
+                        apic_assign_obj.aim_aid_id != self.agent_id and
+                            not apic_assign_obj.is_available(aim_context)):
+                        continue
+                    # This means the original aim-aid owner might have
+                    # crashed or something. We will just take it!
+                    if apic_assign_obj and self.session.ipaddr in url:
+                        obj = self._update_apic_assign_db(
+                            aim_context, apic_assign, apic_assign_obj)
+                        if obj is None:
+                            continue
+                        self.need_recovery = False
+                        self.apic_assign_obj = obj
+                        return
+                    session = self._ws_session_login(url, url_max_retries,
+                                                     RECOVERY_PURPOSE)
+                    if session:
+                        obj = self._update_apic_assign_db(
+                            aim_context, apic_assign, apic_assign_obj)
+                        if obj:
+                            if self.session and self.session.session:
+                                self.session.close()
+                                self.is_session_reconnected = True
+                            self.session = session
+                            self._spawn_monitors()
+                            self.need_recovery = False
+                            self.apic_assign_obj = obj
+                            return
+                        else:
+                            session.close()
+                            continue
```

Comment:
It seems that similar sequences of code are repeated several times. Should this be refactored into another helper function, or maybe moved into _ws_session_login() itself? 

Its also not really clear why _update_apic_assign_db() is called in some cases but not all. This could be parameterized if needed.

Reply:
I think it makes sense to put some of that into _ws_session_login() so will look into it. We don't need to call _update_apic_assign_db() when we try to find a back-up APIC because at this point we are going to share an APIC who already has an owner so no need to claim the ownership of that APIC.